src/controller/CommissionerDiscoveryController.cpp

```diff
@@ -94,24 +120,73 @@ void CommissionerDiscoveryController::OnUserDirectedCommissioningRequest(UDCClie
-                                        RotatingDeviceId::kHexMaxLength);
+    char rotatingIdString[Dnssd::kMaxRotatingIdLen * 2 + 1];
+    Encoding::BytesToUppercaseHexString(state.GetRotatingId(), state.GetRotatingIdLength(), rotatingIdString,
+                                        sizeof(rotatingIdString));
 
     ChipLogDetail(Controller,
                   "------PROMPT USER: %s is requesting permission to cast to this TV, approve? [" ChipLogFormatMEI
                   "," ChipLogFormatMEI ",%s,%s]",
                   state.GetDeviceName(), ChipLogValueMEI(state.GetVendorId()), ChipLogValueMEI(state.GetProductId()),
-                  state.GetInstanceName(), rotatingDeviceIdHexBuffer);
+                  state.GetInstanceName(), rotatingIdString);
     if (mUserPrompter != nullptr)
     {
         mUserPrompter->PromptForCommissionOKPermission(state.GetVendorId(), state.GetProductId(), state.GetDeviceName());
     }
     ChipLogDetail(Controller, "------Via Shell Enter: controller ux ok|cancel");
 }
 
+/// Callback for getting execution into the main chip thread
+void CallbackOk(System::Layer * aSystemLayer, void * aAppState)
+{
+    ChipLogError(AppServer, "UX Ok: now on main thread");
+    GetCommissionerDiscoveryController()->InternalOk();
+}
+
+/// Callback for getting execution out of the main chip thread
+void CallbackExternalWorkerOk(System::Layer * aSystemLayer, void * aAppState)
+{
+    ChipLogError(AppServer, "UX Ok: out of main thread");
+
+    UDCClientState * client = static_cast<UDCClientState *>(aAppState);
+
+    char rotatingIdString[Dnssd::kMaxRotatingIdLen * 2 + 1] = "";
+    Encoding::BytesToUppercaseHexString(client->GetRotatingId(), client->GetRotatingIdLength(), rotatingIdString,
+                                        sizeof(rotatingIdString));
+    CharSpan rotatingIdSpan = CharSpan(rotatingIdString, sizeof(rotatingIdString));
+
+    uint8_t targetAppCount = client->GetNumTargetAppInfos();
+    if (targetAppCount > 0)
+    {
+        for (uint8_t i = 0; i < targetAppCount; i++)
+        {
+            TargetAppInfo info;
+            if (client->GetTargetAppInfo(i, info))
+            {
+                GetCommissionerDiscoveryController()->GetPasscodeService()->LookupTargetContentApp(
+                    client->GetVendorId(), client->GetProductId(), rotatingIdSpan, info);
+            }
+        }
+        return;
+    }
+
+    GetCommissionerDiscoveryController()->GetPasscodeService()->FetchCommissionPasscodeFromContentApp(
+        client->GetVendorId(), client->GetProductId(), rotatingIdSpan);
+}
+
 void CommissionerDiscoveryController::Ok()
 {
+    ChipLogError(AppServer, "UX Ok: moving to main thread");
+    // need to ensure callback is on main chip thread
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackOk, nullptr);
+}
+
+void CommissionerDiscoveryController::InternalOk()
+{
+    ChipLogError(AppServer, "UX InternalOk");
```

Comment:
Should `assertChipStackLockedByCurrentThread()`, yes?  And include the right header for that.

Reply:
addressed